File: Simple-Larrabee/Larrabee.m
clear all
%--------------------------------
%	ペラ設定値
%--------------------------------
B = 2;							%ブレード数B
n = 100;						%分割数n
D = 3.2;						%ペラ直径D[m]
R = D / 2.0;					%ペラ半径R（Dから求まる）
dr = D / 2.0 / n;				%ペラ分割幅Δr（Dとnから求まる）
r = 0.1:dr:R;					%ペラ位置r（ベクトル）
rpm = 130;						%回転数[rpm]
Omega = rpm / 60 * 2 * pi;		%回転数Ω[rad/s]（rpmから求まる）
V = 7.2;						%機速[m/s]
rho = 1.184;					%空気密度ρ[kg/m^3]
nu = 1.540 * 10^(-5);			%動粘性係数ν[m^2/s]
T = 30;							%必要推力T[N]

%翼型のデータ。Clを決めて、それに対応するCdとalphaを入力
Cl = 0.8;							%（局所）揚力係数
Cd = 0.015;						%（局所）抗力係数
alpha_deg = 4;					%迎え角α[degree]

%-------------------------------
alpha = alpha_deg .* pi/180;		%迎え角α[rad]
DL = Cd ./ Cl;						%D/L：揚抗比の逆数
Tc = 2*T / (rho*V^2*pi*R^2);		%推力係数Tc

%--------------------------------
%	Betz & Prandtlの式
%--------------------------------
x = Omega .*r / V;					%局所進行率の逆数x(ベクトル）
sinphi = sqrt(1+x.^2).^(-1);		%sinphi :sin(phi)（ベクトル）
cosphi = x .* sqrt(1+x.^2).^(-1);	%cosphi :cos8phi)（ベクトル）
lambda = V / Omega / R;				%進行率λ（スカラー）
f = B/2 * sqrt(lambda^2 +1)/lambda *(1-r/R); %渦面間隔パラメータf(vortex sheet spacing parameter)（ベクトル）
F = 2/pi * acos(e.^(-f));			%Plandtlの渦間隔パラメータF（ベクトル）
G = F.*x.^2 ./ (1+x.^2);			%Betz & Prandtlの最小誘導損失のプロペラの循環関数G（ベクトル）

xi = r/R;							%無次元ペラ位置ξ（ベクトル）
dxi = 1/n;							%Δξ（スカラー）

%--------------------------------
%	Larrabeeの設計法(論文の式順序に従う)
%--------------------------------
%ζを出すための積分式I1,I2
% dI1 = G .* (1 - DL./x) .* xi * dxi;	
% dI2 = G .* (1 - DL./x) .*xi ./ (x.^2+1) * dxi;
% I1 = 4 * sum(dI1);
% I2 = 2 * sum(dI2);
dI1dxi = G .* (1 - DL./x) .* xi;
dI2dxi = G .* (1 - DL./x) .*xi ./ (x.^2+1);
I1 = 4 * trapz(xi,dI1dxi);
I2 = 2 * trapz(xi,dI2dxi);


%渦面移動速度比ζ（スカラー）
zeta = I1 / (2*I2) * (1 - sqrt(1 - (4*I2*Tc / I1^2)));	
%渦面移動速度v'（スカラー）vortex displacement velocity
vd = zeta * V;

%循環分布Γ（(6)式から）
Gamma = 2 * pi .*r * vd .* sinphi .* cosphi .* F / B;
ad = 0.5 * vd / V ./ (x.^2+1);		%a'

%平面図関数（=(c/R)*Cl/zeta）planform function
planform = 4*pi / B * lambda .*G ./ sqrt(1+x.^2);

chord = planform * zeta ./ Cl * R;	%コード長c
phi = atan(lambda ./xi * (1 + zeta/2));	%らせん角度Φ
beta = phi + alpha;					%ピッチ角β[rad]
beta_deg = beta *180/pi;			%ピッチ角β[deg]

dTdrL = rho * Omega .*r .* (1 - ad) * B .* Gamma;	%流入速度による局所推力
dTdr = dTdrL .* (1- DL./x);		%誘導速度考慮した局所推力（抗力方向の誘導速度成分によって局所推力が減少）
% dT = dTdr .*dr;					%ΔT
% T = sum(dT) * B					%推力T ブレード数考慮
T = trapz(r,dTdr)					%推力T 台形公式積分

epsilon = atan(DL);					%揚力vs抗力の角度ε[rad]
etae = tan(phi) ./ tan(phi+epsilon) .* (1/(1+0.5*zeta)); %ブレード要素での効率ηe

dTcdxi = 2 .* zeta * G .*(1-DL./x).*xi.*(2-zeta./(x.^2+1));	%dTc/dξ（(16)式）
% eta = sum(etae .* dTcdxi .*dxi / Tc);	%効率η
eta = trapz(xi,etae .* dTcdxi) /Tc		%効率η
dQdr = dTdr * V ./ (eta * Omega);		%局所トルク[Nm]
Q = T * V / (eta * Omega);				%トルク[Nm]
W = Q * Omega							%必要パワー[W]

Re = sqrt(V^2 + (Omega .*r .* (1-ad)).^2) .* chord /nu;	%レイノルズ数Re

%--------------------------------
% 	グラフ描写と画像書き出し
%--------------------------------

%コード長とピッチ角のグラフ
figure(1);
	subplot(2,1,1);
		plot(r,chord*1000);
		xlabel('r');
		ylabel('chord[mm]');
		xlim([0 R]);
		grid on;

	subplot(2,1,2);
		plot(r,beta_deg);
		xlabel('r');
		ylabel('beta[deg]');
		xlim([0 R]);
		grid on;
	%-r100は解像度を100dpiにする。文字を大きくするために100dpi
	print -dpng -r100 chord_pitch.png;
	
%推力とトルクのグラフ
figure(2)
	plot(r,dTdr,r,dQdr);
	ylabel('Thrust[N] , Torque[Nm]');
	xlim([0 R]);
	legend('Thrust[N]','Torque[Nm]');
	grid on;
	print -dpng -r100 thrust.png

%レイノルズ数のグラフ
figure(3)
	plot(r,Re);
	ylabel('Reinolds number');
	xlim([0 R]);
	grid on;
	print -dpng -r100 Re.png

File: simple_Larrabee_python/Larrabee.py
import numpy as np
import matplotlib.pyplot as plt

# 入力===============================
# プロペラ設計条件
num_blade = 2       # ブレード数B
num_div = 100       # 分割数n
dia = 3.2           # ペラ直径D[m]
rpm = 130           # 回転数[rpm]
vel = 7.2           # 機速[m/s]
thrust = 30         # 必要推力thrust[N]

# 翼型データ。Clを決めて、それに対応するCdとalphaを入力
coef_lift = 0.8     # 局所揚力係数Cl
coef_drag = 0.014   # 局所抗力係数Cd
alpha_deg = 3       # 迎角α[deg]

# 環境条件
rho = 1.184         # 空気密度ρ[kg/m^3]
nu = 1.540e-5       # 動粘性係数ν[m^2/s]
# 入力ここまで===============================

radius = dia / 2.0  # ペラ半径R[m]
dr = dia / 2.0 / num_div    # 分割幅Δr[m]
r = np.arange(0.1, radius + dr, dr)  # ペラ位置r[m]
omega = rpm / 60 * 2 * np.pi    # 回転数Ω[rad/s]
alpha = alpha_deg * np.pi / 180  # 迎角α[rad]
drag_lift_ratio = coef_drag / coef_lift  # 揚抗比L/Dの逆数
Tc = 2 * thrust / (rho * vel**2 * np.pi * radius**2)  # 推力係数Tc

# ================================
# 	Betz & Prandtlの式
# ================================
# --------------------------------
# 	x		:局所進行率の逆数x
# 	sinphi	:sin(phi)
# 	cosphi	:cos8phi)
# 	lambda	:進行率λ（スカラー）
# 	f		:渦面間隔パラメータf(vortex sheet spacing parameter)
# 	F		:Plandtlの渦間隔パラメータF
# 	G		:Betz & Prandtlの最小誘導損失のプロペラの循環関数G

# 	xi		:無次元ペラ位置ξ
# 	dxi		:Δξ（スカラー）
# --------------------------------

# ----論文(3)to(5)
x = omega * r / vel
sinphi = np.sqrt(1 + x**2)**(-1)
cosphi = x * np.sqrt(1 + x**2)**(-1)

# ----論文(6)to(9)
lambda_val = vel / omega / radius
f = num_blade / 2 * np.sqrt(lambda_val**2 + 1) / lambda_val * (1 - r / radius)
F = 2 / np.pi * np.arccos(np.clip(np.exp(-f), -1, 1))
G = F * x**2 / (1 + x**2)

# ----論文(17)
xi = r / radius
dxi = 1 / num_div

# ================================
# 	Larrabeeの設計法(論文の式順序に従う)
# ================================
# --------------------------------
# 	I1,I2	:ζを出すための積分式
# 	dI1dxi	:dI1/dξ
# 	dI2dxi	:dI2/dξ
# 	zeta	:渦面移動速度比ζ（スカラー）
# 	vd		:渦面移動速度v'（スカラー）

# 	Gamma	:循環分布Γ（論文(6)式から）
# 	ad		:a'
# 	planform:平面図関数（=(c/R)*Cl/zeta）planform function
# 	chord	:コード長c
# 	phi		:らせん角度Φ
# 	beta	:ピッチ角β[rad]
# 	beta_deg:ピッチ角β[deg]

# 	dTdrL	:流入速度による局所推力(dT/dr)_L
# 	dTdr	:誘導速度考慮した局所推力（誘導速度成分によって局所推力が減少）dT/dr
# 	dT		:ΔT（使わない)
# 	T		:推力T

# 	epsilon	:揚力vs抗力の角度ε[rad]
# 	etae	:ブレード要素での効率ηe
# 	dTcdxi	:dTc/dξ（論文(16)式）
# 	eta		:効率η(<1)
# 	dQdr	:局所トルクdQ/dr[Nm]
# 	Q		:トルクQ[Nm]
# 	W		:必要パワーW[W]
# 	Re		:レイノルズ数Re
# --------------------------------

# ----論文(18)to(21)
dI1dxi = G * (1 - drag_lift_ratio / x) * xi
dI2dxi = G * (1 - drag_lift_ratio / x) * xi / (x**2 + 1)
I1 = 4 * np.trapz(dI1dxi, xi)
I2 = 2 * np.trapz(dI2dxi, xi)
zeta = I1 / (2 * I2) * (1 - np.sqrt(1 - (4 * I2 * Tc / I1**2)))
vd = zeta * vel

# ----論文(6),(11)
Gamma = 2 * np.pi * r * vd * sinphi * cosphi * F / num_blade
ad = 0.5 * vd / vel / (x**2 + 1)

# ----論文(24)to(26)
planform = 4 * np.pi / num_blade * lambda_val * G / np.sqrt(1 + x**2)
chord = planform * zeta / coef_lift * radius
phi = np.arctan(lambda_val / xi * (1 + zeta / 2))
beta = phi + alpha
beta_deg = beta * 180 / np.pi

# ----論文(12)to(14)
dTdrL = rho * omega * r * (1 - ad) * num_blade * Gamma
dTdr = dTdrL * (1 - drag_lift_ratio / x)
T = np.trapz(dTdr, r)

# ----論文(28),(29),(16)
epsilon = np.arctan(drag_lift_ratio)
etae = np.tan(phi) / np.tan(phi + epsilon) * (1 / (1 + 0.5 * zeta))
dTcdxi = 2*zeta * G * (1 - drag_lift_ratio / x) * xi * (2 - zeta / (x**2 + 1))
eta = np.trapz(etae * dTcdxi, xi) / Tc
dQdr = dTdr * vel / (eta * omega)
Q = T * vel / (eta * omega)
W = Q * omega

Re = np.sqrt(vel**2 + (omega * r * (1 - ad))**2) * chord / nu

# 出力================================
print("---- 入力値 ----")
print(f"ブレード数B = {num_blade} [本]")
print(f"プロペラ直径D = {2 * radius:.2f} [m]")
print(f"プロペラ回転数n = {rpm} [rpm]")
print(f"機体速度V = {vel:.2f} [m/s]")
print(f"空気密度ρ = {rho:.2f} [kg/m^3]")
print(f"動粘性係数μ = {nu:.2e} [m^2/s]")
print(f"必要推力Tc = {thrust:.2f} [N]")
print("---- 計算結果 ----")
print(f"推力T = {T:.2f} [N]")
print(f"トルクQ = {Q:.2f} [Nm]")
print(f"必要パワーW = {W:.1f} [W]")
print(f"効率η = {eta*100:.1f} [%]")

# コード長とピッチ角のグラフ
fig, (ax1, ax2) = plt.subplots(2, 1)
ax1.plot(r, chord * 1000)
ax1.set_ylabel('chord [mm]')
ax1.set_xlim([0, radius])
ax1.grid()

ax2.plot(r, beta_deg)
ax2.set_xlabel('r [m]')
ax2.set_ylabel('beta [deg]')
ax2.set_xlim([0, radius])
ax2.grid()

plt.tight_layout()
fig.savefig('chord_pitch.png')

# 推力とトルクのグラフ
plt.figure()
plt.plot(r, dTdr, label='Thrust [N]')
plt.plot(r, dQdr, label='Torque [Nm]')
plt.xlabel('r [m]')
plt.ylabel('Thrust [N] , Torque [Nm]')
plt.xlim(0, radius)
plt.legend()
plt.grid()
plt.tight_layout()
plt.savefig('thrust.png')

# レイノルズ数のグラフ
plt.figure()
plt.plot(r, Re)
plt.xlabel('r [m]')
plt.ylabel('Reinolds number [-]')
plt.xlim(0, radius)
plt.grid()
plt.tight_layout()
plt.savefig('Re.png')

plt.show()
